File: backend/src/main/kotlin/com/aicoach/service/StravaActivityService.kt
package com.aicoach.service

import com.aicoach.config.StravaConfig
import com.aicoach.domain.dto.ActivityDto
import com.aicoach.domain.dto.SyncResultDto
import com.aicoach.domain.entity.StravaActivity
import com.aicoach.repository.AthleteRepository
import com.aicoach.repository.StravaActivityRepository
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.reactive.function.client.WebClient
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

@Service
class StravaActivityService(
    private val stravaProps: StravaConfig.StravaProperties,
    private val oauthService: StravaOAuthService,
    private val activityRepository: StravaActivityRepository,
    private val athleteRepository: AthleteRepository,
    private val webClientBuilder: WebClient.Builder
) {
    private val log = LoggerFactory.getLogger(StravaActivityService::class.java)
    private val dtFormatter = DateTimeFormatter.ISO_DATE_TIME

    @Transactional
    fun syncActivities(): SyncResultDto {
        val token = oauthService.getValidToken()
            ?: return SyncResultDto(0, "No Strava token found. Please connect Strava first.")

        val client = webClientBuilder
            .baseUrl(stravaProps.apiBaseUrl)
            .defaultHeader("Authorization", "Bearer ${token.accessToken}")
            .build()

        var page = 1
        var totalSynced = 0

        while (true) {
            @Suppress("UNCHECKED_CAST")
            val activities = client.get()
                .uri("/athlete/activities?per_page=100&page=$page")
                .retrieve()
                .bodyToMono(List::class.java)
                .block() as? List<Map<String, Any>> ?: break

            if (activities.isEmpty()) break

            for (activity in activities) {
                val sportType = activity["sport_type"] as? String ?: activity["type"] as? String ?: ""
                if (sportType !in listOf("Run", "TrailRun")) continue
                val stravaId = (activity["id"] as Number).toLong()
                if (!activityRepository.existsByStravaId(stravaId)) {
                    val saved = saveActivity(activity, token.athleteId)
                    if (saved != null) totalSynced++
                }
            }

            if (activities.size < 100) break
            page++
        }

        log.info("Synced $totalSynced new activities from Strava")
        return SyncResultDto(totalSynced, "Successfully synced $totalSynced new activities")
    }

    @Transactional
    fun syncActivitiesForAthlete(internalAthleteId: Long, afterDate: LocalDate? = null): SyncResultDto {
        val token = oauthService.getValidTokenForAthlete(internalAthleteId)
        if (token == null) {
            log.warn("sync: no token found for internal athlete $internalAthleteId")
            return SyncResultDto(0, "No Strava token found for athlete $internalAthleteId. Please connect Strava first.")
        }
        log.info("sync: found token for internal athlete $internalAthleteId (stravaAthleteId=${token.athleteId}, expiresAt=${token.expiresAt}, internalAthlete=${token.internalAthlete?.id})")

        val client = webClientBuilder
            .baseUrl(stravaProps.apiBaseUrl)
            .defaultHeader("Authorization", "Bearer ${token.accessToken}")
            .build()

        val afterEpoch: Long? = afterDate?.atStartOfDay(ZoneOffset.UTC)?.toEpochSecond()
        log.info("sync: afterDate=$afterDate afterEpoch=$afterEpoch")

        var page = 1
        var totalSynced = 0
        var totalFetched = 0
        var totalSkippedType = 0
        var totalAlreadyExists = 0

        while (true) {
            val uri = buildString {
                append("/athlete/activities?per_page=100&page=$page")
                if (afterEpoch != null) append("&after=$afterEpoch")
            }
            log.info("sync: fetching page $page — $uri")

            val rawResponse = client.get()
                .uri(uri)
                .retrieve()
                .bodyToMono(Any::class.java)
                .block()

            log.info("sync: page $page raw response type=${rawResponse?.javaClass?.simpleName}")

            @Suppress("UNCHECKED_CAST")
            val activities = rawResponse as? List<Map<String, Any>>
            if (activities == null) {
                log.warn("sync: page $page did not return a list — response=$rawResponse")
                break
            }

            log.info("sync: page $page returned ${activities.size} activities")
            if (activities.isEmpty()) break
            totalFetched += activities.size

            for (activity in activities) {
                val sportType = activity["sport_type"] as? String ?: activity["type"] as? String ?: ""
                if (sportType !in listOf("Run", "TrailRun")) {
                    totalSkippedType++
                    continue
                }
                val stravaId = (activity["id"] as Number).toLong()
                val existing = activityRepository.findByStravaId(stravaId).orElse(null)
                if (existing == null) {
                    val saved = saveActivityForInternalAthlete(activity, token.athleteId, internalAthleteId)
                    if (saved != null) totalSynced++
                } else if (existing.internalAthlete?.id != internalAthleteId) {
                    existing.internalAthlete = athleteRepository.findById(internalAthleteId).orElse(null)
                    activityRepository.save(existing)
                    totalSynced++
                } else {
                    totalAlreadyExists++
                }
            }

            if (activities.size < 100) break
            page++
        }

        log.info("sync complete: fetched=$totalFetched skippedType=$totalSkippedType alreadyExists=$totalAlreadyExists newlySaved=$totalSynced")
        return SyncResultDto(totalSynced, "Successfully synced $totalSynced new activities")
    }

    fun getActivities(pageable: Pageable): Page<ActivityDto> {
        return activityRepository
            .findBySportTypeInOrderByActivityDateDesc(
                listOf("Run", "TrailRun", "VirtualRun"),
                pageable
            )
            .map { it.toDto() }
    }

    fun getActivitiesForAthlete(internalAthleteId: Long, pageable: Pageable): Page<ActivityDto> {
        return activityRepository
            .findByInternalAthleteIdAndSportTypeInOrderByActivityDateDesc(
                internalAthleteId,
                listOf("Run", "TrailRun", "VirtualRun"),
                pageable
            )
            .map { it.toDto() }
    }

    private fun saveActivity(data: Map<String, Any>, athleteId: Long): StravaActivity? {
        return try {
            val sportType = data["sport_type"] as? String ?: data["type"] as? String ?: "Run"
            val startDateStr = data["start_date_local"] as? String ?: data["start_date"] as? String ?: return null
            val startDt = LocalDateTime.parse(startDateStr.replace("Z", ""), dtFormatter)

            val activity = StravaActivity(
                stravaId = (data["id"] as Number).toLong(),
                athleteId = athleteId,
                name = data["name"] as? String,
                sportType = sportType,
                activityDate = startDt.toLocalDate(),
                startDatetime = startDt,
                distanceM = (data["distance"] as? Number)?.let { BigDecimal(it.toString()) },
                movingTimeS = (data["moving_time"] as? Number)?.toInt(),
                elapsedTimeS = (data["elapsed_time"] as? Number)?.toInt(),
                totalElevationM = (data["total_elevation_gain"] as? Number)?.let { BigDecimal(it.toString()) },
                averageSpeed = (data["average_speed"] as? Number)?.let { BigDecimal(it.toString()) },
                maxSpeed = (data["max_speed"] as? Number)?.let { BigDecimal(it.toString()) },
                averageHeartrate = (data["average_heartrate"] as? Number)?.let { BigDecimal(it.toString()) },
                maxHeartrate = (data["max_heartrate"] as? Number)?.let { BigDecimal(it.toString()) },
                trainer = data["trainer"] as? Boolean ?: false,
                manual = data["manual"] as? Boolean ?: false
            )
            activityRepository.save(activity)
        } catch (e: Exception) {
            log.warn("Failed to save activity ${data["id"]}: ${e.message}")
            null
        }
    }

    private fun saveActivityForInternalAthlete(
        data: Map<String, Any>,
        stravaAthleteId: Long,
        internalAthleteId: Long
    ): StravaActivity? {
        return try {
            val sportType = data["sport_type"] as? String ?: data["type"] as? String ?: "Run"
            val startDateStr = data["start_date_local"] as? String ?: data["start_date"] as? String ?: return null
            val startDt = LocalDateTime.parse(startDateStr.replace("Z", ""), dtFormatter)

            val activity = StravaActivity(
                stravaId = (data["id"] as Number).toLong(),
                athleteId = stravaAthleteId,
                name = data["name"] as? String,
                sportType = sportType,
                activityDate = startDt.toLocalDate(),
                startDatetime = startDt,
                distanceM = (data["distance"] as? Number)?.let { BigDecimal(it.toString()) },
                movingTimeS = (data["moving_time"] as? Number)?.toInt(),
                elapsedTimeS = (data["elapsed_time"] as? Number)?.toInt(),
                totalElevationM = (data["total_elevation_gain"] as? Number)?.let { BigDecimal(it.toString()) },
                averageSpeed = (data["average_speed"] as? Number)?.let { BigDecimal(it.toString()) },
                maxSpeed = (data["max_speed"] as? Number)?.let { BigDecimal(it.toString()) },
                averageHeartrate = (data["average_heartrate"] as? Number)?.let { BigDecimal(it.toString()) },
                maxHeartrate = (data["max_heartrate"] as? Number)?.let { BigDecimal(it.toString()) },
                trainer = data["trainer"] as? Boolean ?: false,
                manual = data["manual"] as? Boolean ?: false
            )
            activity.internalAthlete = athleteRepository.findById(internalAthleteId).orElse(null)
            activityRepository.save(activity)
        } catch (e: Exception) {
            log.warn("Failed to save activity ${data["id"]}: ${e.message}")
            null
        }
    }

    private fun StravaActivity.toDto() = ActivityDto(
        id = id,
        stravaId = stravaId,
        name = name,
        sportType = sportType,
        activityDate = activityDate,
        startDatetime = startDatetime,
        distanceKm = distanceKm,
        movingTimeS = movingTimeS,
        totalElevationM = totalElevationM,
        averageHeartrate = averageHeartrate
    )
}

File: server/src/services/stravaActivityService.ts
import axios from 'axios'
import { db } from '../db/client.js'
import { stravaActivities, athletes } from '../db/schema.js'
import { eq, and, inArray, gte, lte } from 'drizzle-orm'
import { getValidToken, getValidTokenForAthlete } from './stravaOAuthService.js'
import type { ActivityDto, SyncResultDto, PagedResponse } from '../types/index.js'

const STRAVA_API_BASE = 'https://www.strava.com/api/v3'
const RUNNING_SPORT_TYPES = ['Run', 'TrailRun', 'VirtualRun']

function toDistanceKm(distanceM: number | null | undefined): number {
  if (!distanceM) return 0
  return Math.round((distanceM / 1000) * 100) / 100
}

function toDto(row: typeof stravaActivities.$inferSelect): ActivityDto {
  return {
    id: row.id,
    stravaId: row.stravaId,
    name: row.name ?? null,
    sportType: row.sportType ?? null,
    activityDate: row.activityDate,
    startDatetime: row.startDatetime ?? null,
    distanceKm: toDistanceKm(row.distanceM),
    movingTimeS: row.movingTimeS ?? null,
    totalElevationM: row.totalElevationM ?? null,
    averageHeartrate: row.averageHeartrate ?? null,
  }
}

function parseActivity(data: Record<string, unknown>, stravaAthleteId: number, internalAthleteId?: number) {
  const startDateStr =
    (data.start_date_local as string | undefined) ??
    (data.start_date as string | undefined) ?? ''
  const startDt = startDateStr.replace('Z', '')
  const activityDate = startDt.substring(0, 10)

  return {
    stravaId: Number(data.id),
    athleteId: stravaAthleteId,
    internalAthleteId: internalAthleteId ?? null,
    name: (data.name as string | undefined) ?? null,
    sportType: (data.sport_type as string | undefined) ?? (data.type as string | undefined) ?? null,
    activityDate,
    startDatetime: startDt || null,
    distanceM: (data.distance as number | undefined) ?? null,
    movingTimeS: (data.moving_time as number | undefined) ?? null,
    elapsedTimeS: (data.elapsed_time as number | undefined) ?? null,
    totalElevationM: (data.total_elevation_gain as number | undefined) ?? null,
    averageSpeed: (data.average_speed as number | undefined) ?? null,
    maxSpeed: (data.max_speed as number | undefined) ?? null,
    averageHeartrate: (data.average_heartrate as number | undefined) ?? null,
    maxHeartrate: (data.max_heartrate as number | undefined) ?? null,
    trainer: (data.trainer as boolean | undefined) ?? false,
    manual: (data.manual as boolean | undefined) ?? false,
  }
}

export async function syncActivities(): Promise<SyncResultDto> {
  const token = await getValidToken()
  if (!token) {
    return { syncedCount: 0, message: 'No Strava token found. Please connect Strava first.' }
  }

  let page = 1
  let totalSynced = 0

  while (true) {
    const resp = await axios.get<unknown[]>(
      `${STRAVA_API_BASE}/athlete/activities?per_page=100&page=${page}`,
      { headers: { Authorization: `Bearer ${token.accessToken}` } }
    )
    const activities = resp.data as Record<string, unknown>[]
    if (!activities || activities.length === 0) break

    for (const activity of activities) {
      const sportType =
        (activity.sport_type as string | undefined) ??
        (activity.type as string | undefined) ?? ''
      if (!RUNNING_SPORT_TYPES.slice(0, 2).includes(sportType)) continue

      const stravaId = Number(activity.id)
      const existing = db
        .select()
        .from(stravaActivities)
        .where(eq(stravaActivities.stravaId, stravaId))
        .get()
      if (!existing) {
        try {
          db.insert(stravaActivities)
            .values(parseActivity(activity, token.athleteId))
            .run()
          totalSynced++
        } catch {
          // skip duplicates
        }
      }
    }

    if (activities.length < 100) break
    page++
  }

  return { syncedCount: totalSynced, message: `Successfully synced ${totalSynced} new activities` }
}

export async function syncActivitiesForAthlete(
  internalAthleteId: number,
  afterDate?: string
): Promise<SyncResultDto> {
  const token = await getValidTokenForAthlete(internalAthleteId)
  if (!token) {
    return {
      syncedCount: 0,
      message: `No Strava token found for athlete ${internalAthleteId}. Please connect Strava first.`,
    }
  }

  const afterEpoch = afterDate
    ? Math.floor(new Date(afterDate).getTime() / 1000)
    : undefined

  let page = 1
  let totalSynced = 0

  while (true) {
    const url = afterEpoch
      ? `${STRAVA_API_BASE}/athlete/activities?per_page=100&page=${page}&after=${afterEpoch}`
      : `${STRAVA_API_BASE}/athlete/activities?per_page=100&page=${page}`

    const resp = await axios.get<unknown[]>(url, {
      headers: { Authorization: `Bearer ${token.accessToken}` },
    })
    const activities = resp.data as Record<string, unknown>[]
    if (!activities || activities.length === 0) break

    for (const activity of activities) {
      const sportType =
        (activity.sport_type as string | undefined) ??
        (activity.type as string | undefined) ?? ''
      if (!['Run', 'TrailRun'].includes(sportType)) continue

      const stravaId = Number(activity.id)
      const existing = db
        .select()
        .from(stravaActivities)
        .where(eq(stravaActivities.stravaId, stravaId))
        .get()

      if (!existing) {
        try {
          db.insert(stravaActivities)
            .values(parseActivity(activity, token.athleteId, internalAthleteId))
            .run()
          totalSynced++
        } catch {
          // skip
        }
      } else if (existing.internalAthleteId !== internalAthleteId) {
        db.update(stravaActivities)
          .set({ internalAthleteId })
          .where(eq(stravaActivities.id, existing.id))
          .run()
        totalSynced++
      }
    }

    if (activities.length < 100) break
    page++
  }

  return { syncedCount: totalSynced, message: `Successfully synced ${totalSynced} new activities` }
}

export function getActivities(page: number, size: number): PagedResponse<ActivityDto> {
  const all = db
    .select()
    .from(stravaActivities)
    .where(inArray(stravaActivities.sportType, RUNNING_SPORT_TYPES))
    .all()
    .sort((a, b) => b.activityDate.localeCompare(a.activityDate))

  const totalElements = all.length
  const totalPages = Math.ceil(totalElements / size)
  const content = all.slice(page * size, page * size + size).map(toDto)

  return { content, totalElements, totalPages, number: page, size }
}

export function getActivitiesForAthlete(
  internalAthleteId: number,
  page: number,
  size: number
): PagedResponse<ActivityDto> {
  const all = db
    .select()
    .from(stravaActivities)
    .where(
      and(
        eq(stravaActivities.internalAthleteId, internalAthleteId),
        inArray(stravaActivities.sportType, RUNNING_SPORT_TYPES)
      )
    )
    .all()
    .sort((a, b) => b.activityDate.localeCompare(a.activityDate))

  const totalElements = all.length
  const totalPages = Math.ceil(totalElements / size)
  const content = all.slice(page * size, page * size + size).map(toDto)

  return { content, totalElements, totalPages, number: page, size }
}

export function getActivitiesByDateRange(
  internalAthleteId: number,
  startDate: string,
  endDate: string
): typeof stravaActivities.$inferSelect[] {
  return db
    .select()
    .from(stravaActivities)
    .where(
      and(
        eq(stravaActivities.internalAthleteId, internalAthleteId),
        gte(stravaActivities.activityDate, startDate),
        lte(stravaActivities.activityDate, endDate)
      )
    )
    .all()
    .filter((a) => RUNNING_SPORT_TYPES.includes(a.sportType ?? ''))
    .sort((a, b) => a.activityDate.localeCompare(b.activityDate))
}
